Extended "ID Document" Attributes

FabrizioMoggio opened the pull request:
#### What type of PR is this?
* enhancement/feature

#### What this PR does / why we need it:
To improve the document identification

#### Which issue(s) this PR fixes:
https://github.com/camaraproject/KnowYourCustomer/issues/197

FabrizioMoggio committed:
Extended "ID Document" Attributes

https://github.com/camaraproject/KnowYourCustomer/issues/197

FabrizioMoggio commented on code/API_definitions/kyc-match.yaml:
```diff
@@ -247,6 +256,20 @@ components:
+        
+        idCitizenIdentifier:
+          type: string
+          description: persistent, unique identifier assigned to a person by the identity provider (usually a government). It must be formatted as CountryCode/UniqueIdentifier
```
> Regardless of the final decision of including the field or not, this description is problematic because, until now, the idDocument was being used as the persistent and unique identifier. This would change the current behaviour of the API and related APIs like Customer Insights. So, in case we end up including the property, it should be considered as an addition to the already existing behaviour and, in any case, modify it.
> 
@fernandopradocabrillo sorry, I'm not sure to get your comment. what behavior changes if we introduce  idCitizenIdentifier?

fernandopradocabrillo replied:
For example, here in Spain, the persistent number is the idDocument, the DNI, assigned by the goverment when you are born. In this proposal, the "idCitizenIdentifier" is the persistent value and the idDocument is a number that changes everytime a new document is reniewed.
It changes the semantic of the idDocument property

HuubAppelboom commented:
@fernandopradocabrillo @FabrizioMoggio In the Netherlands, the persistent number BSN is heavily regulated, and is not allowed to be used by telcos for identification purposes (Only government and a few private sectors are allowed its processing).  Hence we use always the document number (which I understood is also used in Spain if the user has no NIF).
In the Netherlands, the persistent number is always present on identity cards, driver licenses, and passports, but on all 3 we have document numbers as well.

For some cases (like the one in Poland) you may want to use both the document number and the persistent personal number. One reason to work with document numbers is that it proof that you have seen an identification document (the personal number is often used in many places, so you can get this from other sources as well), wheras the personal number is ofcourse a much better identifier.   Also, document numbers are important in case identity document are lost or stolen. In most countries there are list of document numbers which are not valid anymore (because of being lost/stolen/fraudulent issued).

 In general, I think we should try to find a solution that works for as many markets as possible, and at the same time be precise what the data means in what we are sharing.

HuubAppelboom commented:
@fernandopradocabrillo @FabrizioMoggio 

One possible solution that may work for most markets is to keep it a single data field, but that in the data field you can put multiple number, with tags what these numbers are.

For example, like this in case you have both (or want to have both verified)
idDocument = "personalNumber:ES/12347699E  documentNumber:ES/32321449" 

or in case you only have the persistent number
idDocument = "personalNumber:ES/12347699E" 

or in case you only have the documentnumber
idDocument = "documentNumber:ES/32321449"" 

What do you think?

HuubAppelboom commented:
@FabrizioMoggio @fernandopradocabrillo  BTW, in other countries like france there are also similar restrictions as in the Netherlands: https://www.privacyworld.blog/2020/01/use-of-the-social-security-number-in-france/

HuubAppelboom commented:
@fernandopradocabrillo @FabrizioMoggio  There is also a ARF published for eIDAS2.0. It may also be a good idea to use the same terminology, see for example in the PID rulebook: https://eu-digital-identity-wallet.github.io/eudi-doc-architecture-and-reference-framework/latest/annexes/annex-3/annex-3.01-pid-rulebook/

So in stead of personalNumber, it would become personalAdministrativeNumber
documentNumber can remain as is, they use the same terminology.

Regadrs
Huub

GillesInnov35 commented:
hi all, joining the discussion, at Orange a unique identifier of the provided document associated to the document type (for example national card or passport) is attached to the contract. Persistent/personal number are never stored.
I'm not sure to clearly understand why multiple identifiers **must** be added to the attributes list to be matched. We should take care about not to make complex the design. If needed an array may be the solution but it will break the backward compatibility, so not the solution.
 ```
 "individualIdentification": [
    {
      "identificationType": "passport",
      "identificationId": "51DD1234"
     }
   ]
```
BR
Gilles

FabrizioMoggio commented:
idDocumentType:
type: string
enum: [passport, national_id_card, residence_permit, diplomatic_id, driver_licence, social_security_id, other]
description: Type of the official identity document provided.

idDocumentExpiryDate:
type: string
format: date
description: Expiration date of the identity document (ISO 8601).

idCitizenIdentifier:
type: string
description: persistent, unique identifier assigned to a person by the identity provider (usually a government)

> @fernandopradocabrillo @FabrizioMoggio
> 
> One possible solution that may work for most markets is to keep it a single data field, but that in the data field you can put multiple number, with tags what these numbers are.
> 
> For example, like this in case you have both (or want to have both verified) idDocument = "personalNumber:ES/12347699E documentNumber:ES/32321449"
> 
> or in case you only have the persistent number idDocument = "personalNumber:ES/12347699E"
> 
> or in case you only have the documentnumber idDocument = "documentNumber:ES/32321449""
> 
> What do you think?

That was one of the options I evaluated but in my opinion it carries the same information of the current solution only making the variables more complex to understand for the Developer. Let me try to implement it here below, than we can have a look on both the solutions (we will align the naming with eIDAS2.0 later):

**Current solution:**

idDocument:
type: string
description: Id number associated to the official identity document in the country. It may contain alphanumeric characters. It typically changes when the document is renewed.

idCitizenIdentifier:
type: string
description: persistent, unique identifier assigned to a person by the identity provider (usually a government)

note: The API Consumer con use both independently. 

**Alternative Solution**

idDocument:
type: string
description: this can be the Id number associated to the official identity document in the country. It may contain alphanumeric characters. It typically changes when the document is renewed. Or it can be a persistent, unique identifier assigned to a person by the identity provider (usually a government). To differentiate between the two possibilities this format must be followed:  "personalNumber:ES/12347699E" or "documentNumber:ES/32321449". IdDocument can carry both the information with comma separation (idDocument= personalNumber:ES/12347699E, documentNumber:ES/32321449)

note: @HuubAppelboom if I got your proposal right, this solution only has one variable containing all the possible information .

For sure the "Alternative Solution" text can be improved, nevertheless I don't see how it helps :-)

FabrizioMoggio commented:
Proposal:

for this PR: 

    documentNumber:
      type: string
      description: Id number associated to the official identity document in the country. It may contain alphanumeric characters. It        typically changes when the document is renewed.

    documentType:
      type: string
      enum: [passport, national_id_card, residence_permit, diplomatic_id, driver_licence, social_security_id, other]
      description: Type of the official identity document provided.

    documentExpiryDate:
      type: string
      format: date
      description: Expiration date of the identity document (ISO 8601).


Open a specific discussion on  "permanent identifier" that is actually an independent topic.

FabrizioMoggio committed:
removed idCitizenIdentifier

https://github.com/camaraproject/KnowYourCustomer/pull/207#issuecomment-2969550446

GillesInnov35 commented:
thanks @FabrizioMoggio , we must not modify the identifier attribute name because already in use (kyc-fill-in also concerned). 
we could propose bellow names, WDYT ?

```
idDocument:
  type: string
  description: Id number associated to the official identity document in the country. It may contain alphanumeric characters. It        typically changes when the document is renewed.

documentType:
  type: string
  enum: [passport, national_id_card, residence_permit, diplomatic_id, driver_licence, social_security_id, other]
  description: Type of the official identity document provided.

documentExpiryDate:
  type: string
  format: date
  description: Expiration date of the identity document (ISO 8601).

```

FabrizioMoggio committed:
renamed idDocumentNumber to idDocument

https://github.com/camaraproject/KnowYourCustomer/pull/207#issuecomment-2969769409

FabrizioMoggio committed:
typo

FabrizioMoggio committed:
Test for Extended "ID Document" Attributes

https://github.com/camaraproject/KnowYourCustomer/issues/197

FabrizioMoggio commented:
> Hi Fabrizio, Hi Gilles,
> 
> Thanks for the proposal. From my side, I have a slight concern whether the enum of the documenttype attributes is enough or not. Maybe we could use 'other' for document types not included, or we could add new types in the future if needed.
> 
> Also, may I double-check if the proposal will not break the backward compatibility of the kyc-match api, please?
> 
> Many thanks, Toshi

the enum for idDocumentType already has "other" : [ passport, national_id_card, residence_permit, diplomatic_id, driver_licence, social_security_id, **other** ]

About the backward compatibility, I don't see any break. I have introduced new parameters and we also decided not to rename idDcument for backward compatibility reasons.  Anyway please have a look and let me know.

HuubAppelboom commented:
> In my understanding the current implementation is agreed, isn't it?

Yes, I think so

FabrizioMoggio commented:
I'm going to create the PR in the new repo as soon as I fix the MegaLinter tool missing in the new repo

FabrizioMoggio commented:
> Thanks, @FabrizioMoggio .
> 
> Also, we should wait for Telefonica Fernando's opinion.
> 
> Best regards, Toshi

Just to be sure, I suppose we should wait for the approval and merging but I can move forward to create the PR. Right? By the way can you have a look on the PR on the Linting tools (https://github.com/camaraproject/KnowYourCustomerMatch/pull/9) that is a prerequisite for any further PR on the code?

ToshiWakayama-KDDI commented:
Close this PR, as this topic has been re-opened in the new kyc-match repository and has been resolved for Fall'25 meta-release.  (See 2025-09-16 minutes)

- Issue #10 https://github.com/camaraproject/KnowYourCustomerMatch/issues/10
- PR #24 https://github.com/camaraproject/KnowYourCustomerMatch/pull/24
